example/main.cpp

```diff
@@ -1,24 +1,139 @@
 #include <iostream>
 #include <threadpool.hpp>
+#include <random>
+#include <condition_variable>
 
 using namespace maxtek;
 
-void trivialFunc()
+class consumerProducerExample
 {
-    while (1)
-    {
-      std::cout << "Hello from task\n";
-    }
-}
+    public:
+        consumerProducerExample(size_t num_consumer_threads = std::thread::hardware_concurrency(),
+        size_t num_producer_threads = std::thread::hardware_concurrency(),
+        size_t buffer_size = 10, int consumer_thread_sleep = 100,
+        int producer_thread_sleep = 100): num_consumer_threads(num_consumer_threads),
+        num_producer_threads(num_producer_threads), buffer_size(buffer_size),
+        consumer_thread_sleep(consumer_thread_sleep),
+        producer_thread_sleep(producer_thread_sleep)
+        {
+            item_buf.reserve(buffer_size);
+            producer_futures.reserve(num_producer_threads);
+            myThreadPool = new threadpool(num_producer_threads + num_consumer_threads);
+        }
+
+        void run()
+        {
+            buffering_active = 1;
+            for (int i = 0 ; i < num_consumer_threads; ++i)
+            {
+                auto consumerTask = std::bind(&consumerProducerExample::consumerThread, this);
+                myThreadPool->submit(consumerTask);
+                // Why didn't the below line work?
+                // myThreadPool->submit(consumerProducerExample::consumerThread);
+            }
+
+            for (int i = 0; i < num_producer_threads; ++i)
+            {
+                auto producerTask = std::bind(&consumerProducerExample::producerThread, this);
+                producer_futures[i] = myThreadPool->submit(producerTask);
+                //myThreadPool->submit(consumerProducerExample::producerThread);
+            }
+        }
+
+        void stop()
+        {
+            std::unique_lock<std::mutex> lock(mtx);
+            buffering_active = 0;
+            lock.unlock();
+            cv.notify_all();
+            for (int i = 0; i < num_producer_threads; ++i)
+            {
+                std::cout << "Producer: " << (i + 1) << ", Num bytes: " << producer_futures[i].get() << "\n";
+            }
+            myThreadPool->shutdown();
+            delete myThreadPool;
+        }
+
+        ~consumerProducerExample()
+        {
+            if (buffering_active)
+            {
+                stop();
+            }
+        }
+    private:
+        size_t num_consumer_threads;
+        size_t num_producer_threads;
+        size_t buffer_size;
+        // in ms
+        int consumer_thread_sleep;
+        // in ms
+        int producer_thread_sleep;
+        bool buffering_active;
+        std::mutex mtx;
+        std::condition_variable cv;
+        std::vector<int> item_buf;
+        threadpool *myThreadPool;
```

Comment:
Make this a `unique_ptr` or `shared_ptr`. We don't need to explicitly demonstrate the destructor by calling `delete`, the auto ptr classes will call it for us.